README.md
<h1 align="center">RealDebrid-Compatible Cloud Storage Mounter for Magisk (prev: rclone-mount)</h1>

<div align="center">
  <!-- Version -->
    <img src="https://img.shields.io/github/v/release/rabbits-heart/rclone_RD-MagiskModule?color=teal&style=for-the-badge"
      alt="Version" />
  <!-- Release Date -->
    <img src="https://img.shields.io/github/release-date/rabbits-heart/rclone_RD-MagiskModule?color=pink&style=for-the-badge"
      alt="_time_stamp_" />
  <!-- Min Magisk -->
    <img src="https://img.shields.io/badge/Magisk-20.4%2B-magenta?&style=for-the-badge"
      alt="_time_stamp_" /></div>

<div align="center">
  <strong>This is the spiritual successor of <a href="https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone">rclone-mount</a> with bug fixes and newer binaries included. More details in the
    <a href="https://github.com/AvinashReddy3108/rclone-mount-magisk/wiki">wiki</a>.
</div>

<div align="center">
  <h3>
    <a href="https://github.com/AvinashReddy3108/rclone-mount-magisk">
      Source Code
    </a>
    <span> | </span>
    <a href="https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone">
      Original Module Repository
    </a>
    <span> | </span>
    <a href="https://github.com/rabbits-heart/rclone_RD-MagiskModule/issues">
      Issues
    </a>
  </h3>
</div>

### About
- This magisk module has been created to work with [@itsToggle](https://github.com/itsToggle)'s [Rclone_RD](https://github.com/itsToggle/rclone_RD/) which implements Real-Debrid to rclone. 

- All credit goes to [@itsToggle](https://github.com/itsToggle), [@Howard20181](https://github.com/Howard20181), [@AvinashReddy3108](https://github.com/AvinashReddy3108) and [@piyushgarg](https://github.com/piyushgarg).

# RClone_RD

This RClone Fork contains a Real-Debrid implementation.
Using this version, the entire RealDebrid /torrents directory can be served as a read-only virtual drive. 

A potential use-case for this is serving the /torrent directory over plex, allowing you to build a media library truly unlimted in size. Im working on a project that allows plex to function the same way that Wako,Syncler and other streaming apps do. Check it out on https://github.com/itsToggle/plex_rd

### Capabilities and Limitations:

- Read/Write capabilities are limited to reading files and deleting them. 
- This rclone fork will automatically sort your torrents into 3 subfolder: "shows", "movies" and "default". If a torrent couldnt be classified as a movie or a show, you can find it in the "default" folder.
- There are no server-side traffic or storage limitations.
- This rclone fork will automatically re-activate direct links when they expire after 1 week.
- There is a server-side connection limit, which I believe is 16 parallel connections.

# Rclone Magisk Module

### Features
- Support for arm, arm64, & x86 (64bit too)
- Huge list of [supported cloud storage providers](https://rclone.org/#providers)
- Apps with ability to specify paths can access the remotes at `/mnt/cloud/`.
- Most file explorers work just fine ([read more on this](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/9)).
- Mount points use names of remote(s) in rclone.conf
- Specify custom rclone params for each remote via `/sdcard/.rclone/.REMOTE.param`
- Access remotes via HTTP or (S)FTP clients, or bind the remotes to `/sdcard/Cloud/REMOTE` (recommended to [read this](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/5)).
- Support for Work Profiles.

### Configuration
1. Copy your `rclone.conf` file (if you have one already) to `/sdcard/.rclone/rclone.conf` (can always be generated later.)
2. Add custom params at `/sdcard/.rclone/.[global/REMOTE].param` (if needed)
3. Install the module via Magisk Manager
4. Run `rclone config` via term if additional setup required
4. All your rclone mount points will show up under `/mnt/cloud/` & `/storage/cloud/` or `/sdcard/cloud/`

For more detailed configuration of rclone please refer to [official documentation](https://rclone.org)

### Known Issues
- VLC  takes a long time to load media as it opens file in write mode when using it's internal browser.

   a. Create remote type alias for media dirs in rclone.conf and
specify `CACHEMODE=off` in `/sdcard/.rclone/.ALIASNAME.param`

- Encrypted devices can not mount until unlock
- Encrypted `rclone.conf` causes reboots
- High cpu/mem in some apps with storage perms ([issue #9](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/9))
- The `fusermount` bin may not be compatible on all devices (see [thread](https://www.google.com/amp/s/forum.xda-developers.com/android/development/fusermount-android-rclone-mount-t3866652/amp/))

## Disclaimer
- Neither the author nor devs will be held responsible for any damage/data loss that may occur during use of this module.
- While we have done our best to make sure no harm will come about, no guarantees can be made.
- Keep in mind the binaries included in this project are BETA quality (at best), which may cause unforseen issues.

Always check this document before updating to new releases as significant changes may occur.

## Credits
- [@itsToggle](https://github.com/itsToggle) for [rclone_RD](https://github.com/itsToggle/rclone_RD/) 
- [@Howard20181](https://github.com/Howard20181), [@AvinashReddy3108](https://github.com/AvinashReddy3108) and [@piyushgarg](https://github.com/piyushgarg) for their original magisk modules
- rclone devs
- pmj_pedro[@xda](https://forum.xda-developers.com/showpost.php?p=78147335&postcount=1)
- agnostic-apollo[@xda](https://forum.xda-developers.com/showpost.php?p=79929083&postcount=12)
- Termux for building and hosting binaries for [rclone](https://packages.termux.org/apt/termux-main/pool/main/r/rclone), [fusermount](https://grimler.se/termux-root-packages-24/pool/stable/libf/libfuse2/), [inotifywait](https://packages.termux.org/apt/termux-main/pool/main/i/inotify-tools), [libandroid-support.so](https://packages.termux.org/apt/termux-main/pool/main/liba/libandroid-support).
- improvements by geofferey@github
- @Zackptg5 for MMT-EX Module template.


CHANGELOG.md
## Changelog (forked)
### v1.30
* Updated `rclone` to match itsToggle fork https://github.com/itsToggle/rclone_RD/releases


### v1.20 (Howard20 fork)
* Updated `rclone` to 1.60.0
* Updated `inotifywait` to 3.22.6.0
* Fix `rclone config` on Windows Subsystem for Android
* Fix bind-mount to `/sdcard` on Windows Subsystem for Android
### v1.14 (binary updates)
* Updated rclone to 1.56.0 for all supported architectures.
* Updated README.md to point to Termux packages source.

### v1.14
* Used @Zackptg5's MMT-EX module template.
* Cleaned all install & wrapper scripts.
* Fixed bugs related to `rclone` command not working (proper wrapper script placement).
* All the webservers (SFTP/FTP/HTTP) work fine now.
* Bring back x86 and x86_64 support.
* Updated all binaries (extracted from Termux builds).
  - `libandroid-support` (28)
  - `rclone` (1.55.0)
  - `fusermount` (2.9.9)
  - `inotifywait` (3.20.11.0)

## Changelog (source)

### v1.13
* Add arm/arm64 1.53 binaries downloaded from https://beta.rclone.org/v1.53.0/testbuilds/rclone-android-16-arm.gz

### v1.12
* Fixed restart problems.

### v1.11
* Add arm/arm64 1.52 bins downloaded from https://beta.rclone.org/v1.52.0/testbuilds/rclone-android-16-arm.gz
* Fixed service.sh paths

### v1.10
* fixed fusermount wrapper

### v1.9
* Add arm/arm64 1.51 bins downloaded from https://beta.rclone.org/
* Commented fusermount wrapper

### v1.8
* Support for Work Profiles `PROFILE=`
* Isolate to Work Profiles `ISOLATE=1`
* Support syncing from SD to remote 

### v1.7
* Add ability to disable HTTP/FTP
* Link rest of default params to custom vars
* Exclude some custom params from globals
* Make some globals exclusive
* Change `BINDPOINT=` to `SDBINDPOINT=`
* Fix bug with custom params
* Set `PATH=` to change priority of used bins

### v1.6
* Simplify custom global parameters
* Fix & improve binding to SD
* Specify additional  rclone ops with `ADD_PARAMS=`
* Replace `rclone mount` ops via `REPLACE_PARAMS=`

### v1.5
* Replace arm/arm64  `rclone` 1.48 bins built with Termux
* Replace arm/arm64 `fusermount` built with Termux
* Add arm/arm64 `libandroid-support.so` from Termux
* Support for mounting to SD
* Squash missing rclone.conf install bug
* Tune default parameters
* Include a wrap for `rclone config`
* Include `fusermount-wrapper.sh`
* General Improvements

### v1.4
* Add ability to disable a remote 
* Add a wrapper script for rclone
* Access remotes via http & ftp
* Use without rebooting device
* Add wrapper cmds to `rclone help`
* Make remount possible via `su -M -c`

### v1.3
* Move user rclone.conf & related to `/sdcard/.rclone/`
* Control global `--vfs-cache-mode` via simple files placed in `/sdcard/.rclone/`
* Specify custom params for individual remotes via `/sdcard/.rclone/.REMOTENAME.params`

### v1.2
* Change install process
* Changes for full systemless
* Improve mount reliability
* Symlink mountpoint to `/storage/`

### v1.1
* Initial release
* rclone mount


syncd.sh
#!/system/bin/sh

MODDIR=${0%/*}
TMPDIR=${MODDIR}/.tmp
SYNC_PENDING=$TMPDIR/$remote.sync.pend

dump_battery() {
    BATTERY_DUMP="$(dumpsys battery)"
}

battery_level() {
    echo "$BATTERY_DUMP" | grep level | cut -d ':' -f2 | cut -d ' ' -f2
}

ac_charge() {
    echo "$BATTERY_DUMP" | grep -w "AC powered" | cut -d ":" -f2 | cut -d " " -f2
}

usb_charge() {
    echo "$BATTERY_DUMP" | grep -w "USB powered" | cut -d ":" -f2 | cut -d " " -f2
}

echo $$ >>"$PIDFILE"

while true; do
    if [ ! -f "$SYNC_PENDING" ]; then
        inotifywait "/storage/emulated/$PROFILE/$SYNCDIR" -e modify,create,moved_to,close_write -q >>/dev/null 2>&1 && touch "$SYNC_PENDING"
    fi

    while true; do
        sleep 5
        dump_battery
        if [ "$(battery_level)" -gt "$SYNC_BATTLVL" ] || [ "$(bettery_level)" -eq "$SYNC_BATTLVL" ] || [ "$(ac_charge)" = true ] || [ "$(usb_charge)" = true ]; then
            echo "Sync battery check success"
        else
            sleep 300
            continue
        fi
        if [ "$SYNC_CHARGE" = 1 ]; then
            if [ "$(ac_charge)" = true ] || [ "$(usb_charge)" = true ]; then
                echo "Sync charge check success"
            else
                echo "Sync charge check fail"
                sleep 300
                continue
            fi
        fi
        if [ "$SYNC_WIFI" = 1 ]; then
            if ! ping -I wlan0 -c 1 "$NETCHK_ADDR" >>/dev/null 2>&1; then
                echo "Sync wifi check fail"
                sleep 300
                continue
            else
                echo "Sync wifi check success"
            fi
        fi
        break
    done

    echo "Syncing..."
    nice -n 19 ionice -c 2 -n 7 "$MODDIR"/rclone copy "/storage/emulated/$PROFILE/$SYNCDIR" "$CLOUDROOTMOUNTPOINT/$remote/$SYNCDIR" --retries-sleep=10m --retries 6 --transfers 1 --multi-thread-streams 1 >>/dev/null 2>&1

    if [ -f "$SYNC_PENDING" ]; then
        rm -f "${SYNC_PENDING:?}"
    fi
    echo "Sync finished!"
done


customize.sh
# shellcheck disable=SC2034
SKIPUNZIP=1
extract() {
    zip=$1
    file=$2
    dir=$3
    junk_paths=$4
    [ -z "$junk_paths" ] && junk_paths=false
    opts="-o"
    [ $junk_paths = true ] && opts="-oj"

    file_path=""
    if [ $junk_paths = true ]; then
        file_path="$dir/$(basename "$file")"
    else
        file_path="$dir/$file"
    fi
    unzip $opts "$zip" "$file" -d "$dir" >&2
    [ -f "$file_path" ] || abort "$file not exists"

    ui_print "- Unzip $file" >&1
}
extract "$ZIPFILE" 'module.prop' "$MODPATH"
extract "$ZIPFILE" 'service.sh' "$MODPATH"
extract "$ZIPFILE" 'uninstall.sh' "$MODPATH"
extract "$ZIPFILE" "syncd.sh" "$MODPATH"
mkdir -p "$MODPATH/system/bin"
extract "$ZIPFILE" "common/binary/$ARCH/inotifywait" "$MODPATH/system/bin" true
extract "$ZIPFILE" "system/bin/rclone" "$MODPATH"
extract "$ZIPFILE" "common/binary/$ARCH/rclone" "$MODPATH" true

if [ "$IS64BIT" = true ]; then
    mkdir -p "$MODPATH/system/lib64"
    extract "$ZIPFILE" "common/lib/$ARCH/libandroid-support.so" "$MODPATH/system/lib64" true
else
    mkdir -p "$MODPATH/system/lib"
    extract "$ZIPFILE" "common/lib/$ARCH/libandroid-support.so" "$MODPATH/system/lib" true
fi

extract "$ZIPFILE" "common/binary/$ARCH/fusermount" "$MODPATH/system/bin" true

set_perm_recursive "$MODPATH" 0 0 0755 0644
set_perm "$MODPATH"/service.sh 0 0 0755
set_perm "$MODPATH"/syncd.sh 0 0 0755
set_perm "$MODPATH"/rclone 0 0 0755
[ -d "$MODPATH/system/bin" ] && set_perm_recursive "$MODPATH/system/bin" 0 0 0755 0755
[ -d "$MODPATH/system/lib" ] && set_perm_recursive "$MODPATH/system/lib" 0 0 0755 0644 u:object_r:system_lib_file:s0
[ -d "$MODPATH/system/lib64" ] && set_perm_recursive "$MODPATH/system/lib64" 0 0 0755 0644 u:object_r:system_lib_file:s0
USER_CONFDIR_OLD=/sdcard/.rclone
USER_CONFDIR=/sdcard/Android/.rclone
USER_CONF=$USER_CONFDIR/rclone.conf
USER_CONF_OLD=$USER_CONFDIR_OLD/rclone.conf
if [ -f "$USER_CONF" ]; then
    ui_print
    ui_print "- Please reboot to take effect"
    ui_print
elif [ -f "$USER_CONF_OLD" ] && [ ! -d "$USER_CONFDIR" ]; then
    ui_print "- Migrating config file to $USER_CONFDIR"
    mkdir -p "$USER_CONFDIR"
    mv "$USER_CONFDIR_OLD"/* "$USER_CONFDIR"/
    rm -rf "${USER_CONFDIR_OLD:?}"
else
    ui_print "'$USER_CONF' not found!"
    ui_print
    ui_print "Additional setup required..."
    ui_print "------------------------------------"
    ui_print " Instructions:                      "
    ui_print " - Open Terminal                    "
    ui_print " - Type 'su' & tap enter            "
    ui_print " - Type 'rclone config' & tap enter "
    ui_print " - Follow on screen options.        "
    ui_print "------------------------------------"
fi


service.sh
#!/system/bin/sh
MODDIR=${0%/*}
# Do NOT assume where your module will be located.
# ALWAYS use $MODDIR if you need to know where this script
# and module is placed.
# This will make sure your module will work
# if Magisk changes it's mount point in the future

# MODULE VARS
CLOUDROOTMOUNTPOINT=/mnt/cloud
USER_CONFDIR=/sdcard/Android/.rclone
USER_CONF=$USER_CONFDIR/rclone.conf
PROFILE=0
DATA_MEDIA=/data/media
RUNTIME_R=/mnt/runtime/read
RUNTIME_W=/mnt/runtime/write
RUNTIME_D=/mnt/runtime/default
BINDPOINT_R=${RUNTIME_R}/emulated/${PROFILE}/Cloud
BINDPOINT_W=${RUNTIME_W}/emulated/${PROFILE}/Cloud
BINDPOINT_D=${RUNTIME_D}/emulated/${PROFILE}/Cloud
BINDPOINT_P=/mnt/pass_through/${PROFILE}/emulated/${PROFILE}/Cloud
SD_BINDPOINT=${BINDPOINT_P}
DISABLE=0
NETCHK=1

# DEFAULT RCLONE PARAMETERS
CONFIGFILE=${MODDIR}/.config/rclone/rclone.conf
LOGFILE=${USER_CONFDIR}/rclone.log
LOGLEVEL=NOTICE
CACHE=/data/rclone/cache
CACHE_BACKEND=/data/rclone/cache/cache-backend
CACHEMODE=off
READCHUNKSIZE=1M
CACHEMAXSIZE=1G
CHUNKSIZE=1M
CHUNKTOTAL=1G
CACHEWORKERS=1
CACHEINFOAGE=1h0m0s
DIRCACHETIME=30m0s
ATTRTIMEOUT=30s
BUFFERSIZE=0
READAHEAD=128k
M_UID=0
M_GID=1015
DIRPERMS=0775
FILEPERMS=0644
UMASK=002
BINDSD=1
SYNC_WIFI=1
SYNC_CHARGE=0
SYNC_BATTLVL=0
HTTP=0
HTTP_ADDR=127.0.0.1:38762
FTP=0
FTP_ADDR=127.0.0.1:38763
SFTP=0
SFTP_ADDR=127.0.0.1:38722
SFTP_USER=
SFTP_PASS=

if [[ -z ${INTERACTIVE} ]]; then
    INTERACTIVE=0
fi

if [[ ! -d ${MODDIR}/.config/rclone ]]; then
    mkdir -p ${MODDIR}/.config/rclone
fi
gethostip() {
    ip=$(ip route list match 0 table all scope global | cut -F3)
    if [ -z "$ip" ]; then
        return 1
    else
        echo "$ip"
    fi
}

while ! (gethostip); do
    sleep 3
done
NETCHK_ADDR=$(gethostip)
custom_params() {
    if [[ ${remote} = global ]]; then
        PARAMS="DISABLE LOGFILE LOGLEVEL CACHEMODE CHUNKSIZE CHUNKTOTAL CACHEWORKERS CACHEINFOAGE DIRCACHETIME ATTRTIMEOUT BUFFERSIZE READAHEAD M_UID M_GID DIRPERMS FILEPERMS READONLY BINDSD ADD_PARAMS REPLACE_PARAMS NETCHK NETCHK_IF NETCHK_ADDR HTTP FTP HTTP_ADDR FTP_ADDR SFTP SFTP_ADDR SFTP_USER SFTP_PASS PROFILE ISOLATE"
    else
        PARAMS="DISABLE LOGFILE LOGLEVEL CACHEMODE CHUNKSIZE CHUNKTOTAL CACHEWORKERS CACHEINFOAGE DIRCACHETIME ATTRTIMEOUT BUFFERSIZE READAHEAD M_UID M_GID DIRPERMS FILEPERMS READONLY BINDSD SDBINDPOINT ADD_PARAMS REPLACE_PARAMS PROFILE ISOLATE SDSYNCDIRS SYNC_WIFI SYNC_BATTLVL SYNC_CHARGE SUBPATH"
    fi
    BAD_SYNTAX="(^\s*#|^\s*$|^\s*[a-z_][^[:space:]]*=[^;&\(\`]*$)"
    if [[ -e ${USER_CONFDIR}/.${remote}.param ]]; then
        echo "Found .${remote}.param"
        if ! [[ $(egrep -q -iv "${BAD_SYNTAX}" ${USER_CONFDIR}/.${remote}.param) ]]; then
            echo "loading .${remote}.param"
            # FIX: Unnecessary and very inefficient double loop
            for PARAM in ${PARAMS[@]}; do
                while read -r VAR; do
                    if [[ "$(echo "${VAR}" | grep "$PARAM=")" ]]; then
                        echo "Importing ${VAR}"
                        VALUE="$(echo ${VAR} | cut -d '=' -f2)"
                        VALUE=\"${VALUE}\"
                        # Unnecessary echo in a subshell execution below? Why not just:
                        # eval "${PARAM}""=""${VALUE}"
                        eval $(echo "${PARAM}""=""${VALUE}")
                    fi
                done <${USER_CONFDIR}/.${remote}.param
            done
        else
            echo ".${remote}.param contains bad syntax"
        fi
    fi
}

global_params() {
    remote=global
    custom_params
    unset remote
    echo
}

net_chk() {
    if [ -z ${NETCHK_IF} ]; then
        NETCHK_IF=" "
    else
        NETCHK_IF=" -I ${NETCHK_IF} "
    fi
    ping ${NETCHK_IF} -c 5 ${NETCHK_ADDR}
}

sd_unbind() {
    if [[ -z ${SDBINDPOINT} ]]; then
        UNBINDPOINT=${BINDPOINT_D}/${remote}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
        UNBINDPOINT=${BINDPOINT_R}/${remote}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
        UNBINDPOINT=${BINDPOINT_W}/${remote}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
        UNBINDPOINT="$SD_BINDPOINT/$remote"
        su -M -c umount -lf "$UNBINDPOINT" >>/dev/null 2>&1
    else
        USER_BINDPOINT=${SDBINDPOINT}
        UNBINDPOINT=${RUNTIME_D}/emulated/${PROFILE}/${USER_BINDPOINT}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
        UNBINDPOINT=${RUNTIME_R}/emulated/${PROFILE}/${USER_BINDPOINT}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
        UNBINDPOINT=${RUNTIME_W}/emulated/${PROFILE}/${USER_BINDPOINT}
        su -M -c umount -lf ${UNBINDPOINT} >>/dev/null 2>&1
    fi
}

sd_binder() {
    if [[ -d ${RUNTIME_D} ]] && [[ ${BINDSD} = 1 ]]; then
        if [[ -z ${SDBINDPOINT} ]]; then
            mkdir -p ${DATA_MEDIA}/${PROFILE}/Cloud/${remote}
            chown media_rw:media_rw ${DATA_MEDIA}/${PROFILE}/Cloud/$remote

            BINDPOINT=${BINDPOINT_D}/${remote}

            su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1

            BINDPOINT=${BINDPOINT_R}/${remote}

            if ! mount | grep -q ${BINDPOINT}; then
                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1
            fi

            BINDPOINT=${BINDPOINT_W}/${remote}

            if ! mount | grep -q ${BINDPOINT}; then
                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1
            fi

            BINDPOINT="$SD_BINDPOINT/$remote"

            if ! mount | grep -q "$BINDPOINT"; then
                su -M -c mount --bind "$CLOUDROOTMOUNTPOINT/$remote" "$BINDPOINT" >>/dev/null 2>&1
            fi

            echo "[$remote] available at: -> [/sdcard/Cloud/${remote}] BINDPOINT=$BINDPOINT"
        else
            mkdir ${DATA_MEDIA}/${PROFILE}/${SDBINDPOINT} >>/dev/null 2>&1
            chown media_rw:media_rw ${DATA_MEDIA}/${PROFILE}/${SDBINDPOINT}

            USER_BINDPOINT=${SDBINDPOINT}
            BINDPOINT=${RUNTIME_D}/emulated/${PROFILE}/${USER_BINDPOINT}

            su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1

            BINDPOINT=${RUNTIME_R}/emulated/${PROFILE}/${USER_BINDPOINT}

            if ! mount | grep -q ${BINDPOINT}; then
                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1
            fi

            BINDPOINT=${RUNTIME_W}/emulated/${PROFILE}/${USER_BINDPOINT}

            if ! mount | grep -q ${BINDPOINT}; then
                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >>/dev/null 2>&1
            fi
            echo "[$remote] available at: -> [/storage/emulated/${PROFILE}/${SDBINDPOINT}]"
            unset BINDPOINT
        fi
    fi
}

syncd_service() {
    if [[ ! -z ${SDSYNCDIRS} ]]; then
        export PIDFILE=${MODDIR}/.tmp/${remote}-syncd.pids
        kill -9 $(cat ${PIDFILE}) >>/dev/null 2>&1
        rm ${PIDFILE} >>/dev/null 2>&1

        if [[ ! -d ${MODDIR}/.tmp ]]; then
            mkdir -p ${MODDIR}/.tmp
        fi

        export CLOUDROOTMOUNTPOINT
        export PROFILE
        export MODDIR
        export remote
        export SYNCWIFI
        export SYNC_BATTLVL
        export SYNC_CHARGE
        export NETCHK_ADDR

        IFS=$':'
        for SYNCDIR in ${SDSYNCDIRS[@]}; do
            export SYNCDIR
            ${MODDIR}/syncd.sh >>/dev/null 2>&1 &
        done
    fi

    unset IFS
    unset SDSYNCDIRS
    unset SYNCDIR
    SYNC_BATTLVL=0
    SYNC_WIFI=1
    SYNC_CHARGE=0
}

reset_params() {
    unset IFS
    unset SDBINDPOINT
    unset BINDSD
    unset ISOLATE
    unset RCLONE_PARAMS
    unset REPLACE_PARAMS
    unset ADD_PARAMS
    unset SYNCDIR
    unset SDSYNCDIRS
    unset PIDFILE
    unset SUBPATH
    LOGFILE=${USER_CONFDIR}/rclone.log
    LOGLEVEL=NOTICE
    CACHEMODE=off
    READCHUNKSIZE=1M
    CACHEMAXSIZE=1G
    CHUNKSIZE=1M
    CHUNKTOTAL=1G
    CACHEWORKERS=1
    CACHEINFOAGE=1h0m0s
    DIRCACHETIME=30m0s
    ATTRTIMEOUT=30s
    BUFFERSIZE=0
    READAHEAD=128k
    M_UID=0
    M_GID=1015
    DIRPERMS=0775
    FILEPERMS=0644
    UMASK=002
    PROFILE=0
    SYNC_WIFI=1
    SYNC_BATTLVL=0
    SYNC_CHARGE=0
}

rclone_mount() {
    if [[ ${READONLY} = 1 ]]; then
        READONLY=" --read-only "
    else
        READONLY=" "
    fi

    if [[ ${ADD_PARAMS} = 0 ]]; then
        unset ADD_PARAMS
    fi

    if [[ ${REPLACE_PARAMS} = 0 ]]; then
        unset REPLACE_PARAMS
    fi

    if [[ -z ${REPLACE_PARAMS} ]]; then
        RCLONE_PARAMS=" --log-file ${LOGFILE} --log-level ${LOGLEVEL} --vfs-cache-mode ${CACHEMODE} --cache-dir ${CACHE} --cache-chunk-path ${CACHE_BACKEND} --cache-db-path ${CACHE_BACKEND} --cache-tmp-upload-path ${CACHE} --vfs-read-chunk-size ${READCHUNKSIZE} --vfs-cache-max-size ${CACHEMAXSIZE} --cache-chunk-size ${CHUNKSIZE} --cache-chunk-total-size ${CHUNKTOTAL} --cache-workers ${CACHEWORKERS} --cache-info-age ${CACHEINFOAGE} --dir-cache-time ${DIRCACHETIME} --attr-timeout ${ATTRTIMEOUT} --cache-chunk-no-memory --use-mmap --buffer-size ${BUFFERSIZE} --max-read-ahead ${READAHEAD} --no-modtime --no-checksum --uid ${M_UID} --gid ${M_GID} --allow-other --dir-perms ${DIRPERMS} --file-perms ${FILEPERMS} --umask ${UMASK} ${READONLY} ${ADD_PARAMS} "
    elif [[ ! -z ${REPLACE_PARAMS} ]]; then
        RCLONE_PARAMS=" ${REPLACE_PARAMS} "
    fi

    if [[ -z ${ADD_PARAMS} ]]; then
        ADD_PARAMS=" "
    elif [[ ! -z ${ADD_PARAMS} ]]; then
        ADD_PARAMS=" ${ADD_PARAMS} "
    fi

    echo "[${remote}] available at: -> [${CLOUDROOTMOUNTPOINT}/${remote}]"
    mkdir -p ${CLOUDROOTMOUNTPOINT}/${remote}
    su -M -p -c nice -n 19 ionice -c 2 -n 7 ${MODDIR}/rclone mount "${remote}:${SUBPATH}" ${CLOUDROOTMOUNTPOINT}/${remote} --config ${CONFIGFILE} ${RCLONE_PARAMS} --daemon >>/dev/null 2>&1 &
}

COUNT=0

if [[ ${INTERACTIVE} = 0 ]]; then
    until [[ $(getprop sys.boot_completed) = 1 ]] && [[ $(getprop dev.bootcomplete) = 1 ]] && [[ $(getprop init.svc.bootanim) = stopped ]] && [[ -e ${USER_CONF} ]] || [[ ${COUNT} -eq 240 ]]; do
        sleep 5
        ((++COUNT))
    done
fi

DECRYPT_CHK() {
    su -M -c ls sdcard | grep -q -w "Android"
}

if [[ ${COUNT} -eq 240 ]] || [[ ! -d /sdcard/Android ]]; then
    exit 0
fi

if [[ ! -d ${USER_CONFDIR} ]]; then
    mkdir ${USER_CONFDIR}
fi

if [[ -e ${USER_CONFDIR}/.disable ]] && [[ ${INTERACTIVE} = 0 ]]; then
    exit 0
fi

if [[ ${INTERACTIVE} = 0 ]]; then
    export INTERACTIVE=1
    ${MODDIR}/service.sh
    exit
fi

global_params

if [ ! -d "$CLOUDROOTMOUNTPOINT" ]; then
    mkdir -p "$CLOUDROOTMOUNTPOINT"
    chown root:sdcard_rw "$CLOUDROOTMOUNTPOINT"
    touch "$CLOUDROOTMOUNTPOINT"/.nomedia
    chown root:sdcard_rw "$CLOUDROOTMOUNTPOINT"/.nomedia
    chmod 0775 "$CLOUDROOTMOUNTPOINT"/.nomedia
fi

if [ ! -f "$CLOUDROOTMOUNTPOINT/.nomedia" ]; then
    touch "$CLOUDROOTMOUNTPOINT"/.nomedia
    chown root:sdcard_rw "$CLOUDROOTMOUNTPOINT"/.nomedia
    chmod 0644 "$CLOUDROOTMOUNTPOINT"/.nomedia
fi

if [[ ! -d ${CACHE} ]]; then
    mkdir -p ${CACHE}
    chown root:sdcard_rw ${CACHE}
    chmod 0775 ${CACHE}
fi

if [[ -d ${CACHE} ]]; then
    chown root:sdcard_rw ${CACHE}
    chmod 0775 ${CACHE}
fi

if [[ ! -d ${CACHE_BACKEND} ]]; then
    mkdir -p ${CACHE_BACKEND}
fi

if [[ -d ${CACHE_BACKEND} ]]; then
    chown root:sdcard_rw ${CACHE_BACKEND}
    chmod 0775 ${CACHE_BACKEND}
fi

if [[ ! -L ${RUNTIME_R}/cloud ]]; then
    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_R}/cloud
fi

if [[ ! -L ${RUNTIME_W}/cloud ]]; then
    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_W}/cloud
fi

if [[ ! -L ${RUNTIME_D}/cloud ]]; then
    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_D}/cloud
fi

if [[ ! -L "$BINDPOINT_P" ]]; then
    ln -sf "$CLOUDROOTMOUNTPOINT" "$BINDPOINT_P"
fi

if [ ! -f "$BINDPOINT_P/.nomedia" ]; then
    touch "$BINDPOINT_P"/.nomedia
    chown root:sdcard_rw "$BINDPOINT_P"/.nomedia
    chmod 0644 "$BINDPOINT_P"/.nomedia
fi

if [[ -e ${USER_CONF} ]]; then
    cp ${USER_CONF} ${CONFIGFILE}
    chmod 0600 ${CONFIGFILE}
fi

if [[ ${NETCHK} = 1 ]]; then
    until net_chk || [[ ${COUNT} = 60 ]]; do
        sleep 5
        ((++COUNT))
    done >>/dev/null 2>&1
fi

echo "Default CACHEMODE ${CACHEMODE}"
sleep 5

LD_LIBRARY_PATH=${MODDIR} ${MODDIR}/rclone listremotes --config ${CONFIGFILE} | cut -f1 -d: |
    while read remote; do
        echo
        list_remote=${remote}
        CLOUDROOTMOUNTPOINT=/mnt/cloud
        PROFILE=0
        DISABLE=0
        READONLY=0
        global_params >>/dev/null 2>&1
        remote=${list_remote}
        custom_params
        if [[ ${ISOLATE} = 1 ]] && [[ ${PROFILE} -gt 0 ]] && [[ ${BINDSD} = 1 ]]; then
            CLOUDROOTMOUNTPOINT=/data/media/${PROFILE}/.cloud
        fi

        if [[ ${DISABLE} = 1 ]] || [[ -e ${USER_CONFDIR}/.${remote}.disable ]]; then
            echo "${remote} disabled by user"
            continue
        fi
        sd_unbind
        rclone_mount
        sd_binder
        syncd_service
        reset_params
    done
echo

if [[ ${HTTP} = 1 ]]; then
    if (${MODDIR}/rclone serve http ${CLOUDROOTMOUNTPOINT} --addr ${HTTP_ADDR} --no-checksum --no-modtime --read-only >>/dev/null 2>&1 &) then
        echo "HTTP Server: http://${HTTP_ADDR}"
    fi
fi

if [[ ${FTP} = 1 ]]; then
    if (${MODDIR}/rclone serve ftp ${CLOUDROOTMOUNTPOINT} --addr ${FTP_ADDR} --no-checksum --no-modtime --read-only >>/dev/null 2>&1 &) then
        echo "FTP Server: ftp://${FTP_ADDR}"
    fi
fi

if [[ ${SFTP} = 1 ]] && [[ ! -z ${SFTP_USER} ]] && [[ ! -z ${SFTP_PASS} ]]; then
    if (${MODDIR}/rclone serve sftp ${CLOUDROOTMOUNTPOINT} --addr ${SFTP_ADDR} --user ${SFTP_USER} --pass ${SFTP_PASS} --no-checksum --no-modtime --read-only >>/dev/null 2>&1 &) then
        echo "SFTP Server: sftp://${SFTP_ADDR}"
    fi
fi

exit
